src/java.base/share/classes/com/sun/crypto/provider/HkdfKeyDerivation.java

```
@@ -0,0 +1,471 @@
+        } catch (NoSuchAlgorithmException nsae) {
+            throw new ProviderException(nsae);
+        }
+
+        if (HKDFTYPE == HKDFTYPES.EXTRACT) {
+            // perform extract
+            try {
+                byte[] extractResult = hkdfExtract(initialKeyMaterial,
+                                                   (salt == null) ? null : salt.getEncoded());
+                return new SecretKeySpec(extractResult, alg);
+            } catch (InvalidKeyException ike) {
+                throw new InvalidParameterSpecException(
+                    "an HKDF Extract could not be initialized with the given key or salt material");
+            }
+
+        } else if (HKDFTYPE == HKDFTYPES.EXPAND) {
+            // perform expand
+            try {
+                byte[] expandResult = hkdfExpand(this.pseudoRandomKey, this.info, this.length);
+                return new SecretKeySpec(expandResult, 0, this.length, alg);
+            } catch (InvalidKeyException ike) {
+                throw new InvalidParameterSpecException(
+                    "an HKDF Expand could not be initialized with the given key material");
+            }
+
+        } else if (HKDFTYPE == HKDFTYPES.EXTRACTEXPAND) {
+            // perform extract and then expand
+            try {
+                byte[] extractResult = hkdfExtract(initialKeyMaterial,
+                                                   (salt == null) ? null : salt.getEncoded());
+                this.pseudoRandomKey = new SecretKeySpec(extractResult, alg);
+                byte[] expandResult = hkdfExpand(this.pseudoRandomKey, this.info, this.length);
+                return new SecretKeySpec(expandResult, 0, this.length, alg);
+            } catch (InvalidKeyException ike) {
+                throw new InvalidParameterSpecException(
+                    "an HKDF ExtractExpand could not be initialized with the given key or salt "
+                    + "material");
+            }
+        }
+
+        return null;
```

Should it throw exception instead of null?